mods/hunger/init.lua

```diff
@@ -0,0 +1,317 @@
+
+HUNGER_MAX = 30			-- maximum level of saturation
+HUNGER_VISUAL_MAX = 20		-- hud bar extends only to 20
+
+
+local function hunger_read(player)
+	local inv = player:get_inventory()
+	if not inv then
+		return nil
+	end
+
+	-- itemstack storage is offest by 1 to make the math work
+	local hgp = inv:get_stack("hunger", 1):get_count()
+	if hgp == 0 then
+		hgp = 21
+		inv:set_stack("hunger", 1, ItemStack({name = ":", count = hgp}))
+	end
+
+	return math.min(HUNGER_MAX + 1, hgp) - 1
+end
+
+local function hunger_save(player)
+	local inv = player:get_inventory()
+	if not inv then
+		return nil
+	end
+	local name = player:get_player_name()
+	local level = hunger_players[name].level
+
+	level = math.min(HUNGER_MAX, level)
+	level = math.max(level, 0)
+
+	inv:set_stack("hunger", 1, ItemStack({name = ":", count = level + 1}))
+	return true
+end
+
+local function hunger_update(player, level)
+	local name = player:get_player_name()
+	if not name then
+		return false
+	end
+	local old = hunger_players[name].level
+	if level == old then
+		return
+	end
+	hunger_players[name].level = level
+
+	player:hud_change(hunger_players[name].hud_id, "number", math.min(HUNGER_VISUAL_MAX, level))
+	hunger_save(player)
+end
+
+local function hunger_node_actions(pos, oldnode, player, ext)
+	if not player or not player:is_player() then
+		return
+	end
+	local name = player:get_player_name()
+	if not name or not hunger_players[name] then
+		return
+	end
+
+	local exhaust = hunger_players[name].exhaust
+	if not exhaust then
+		hunger_players[name].exhaust = 0
+	end
+
+	local new = HUNGER_EXHAUST_PLACE
+
+	-- placenode event
+	if not ext then
+		new = HUNGER_EXHAUST_DIG
+	end
+
+	-- assume its send by action_timer(globalstep)
+	if not pos and not oldnode then
+		new = HUNGER_EXHAUST_MOVE
+	end
+
+	exhaust = exhaust + new
+
+	if exhaust > HUNGER_EXHAUST_LVL then
+		exhaust = 0
+		local h = tonumber(hunger_players[name].level)
+		if h > 0 then
+			hunger_update(player, h - 1)
+		end
+	end
+
+	hunger_players[name].exhaust = exhaust
+end
+
+local function hunger_craft_actions(itemstack, player, old_craft_grid, craft_inv)
+	if not player or not player:is_player() then
+		return
+	end
+
+	local name = player:get_player_name()
+	if not name or not hunger_players[name] then
+		return
+	end
+
+	local exhaust = hunger_players[name].exhaust
+	if not exhaust then
+		hunger_players[name].exhaust = 0
+		return
+	end
+
+	if exhaust > HUNGER_EXHAUST_LVL then
+		exhaust = 0
+		local h = tonumber(hunger_players[name].level)
+		if h > 0 then
+			hunger_update(player, h - 1)
+		end
+	end
+
+	hunger_players[name].exhaust = exhaust
+end
+
+-- TODO: add hunger for fighting
+
+-- Time based hunger functions
+local hunger_timer = 0
+local health_timer = 0
+local action_timer = 0
+
+local function hunger_globaltimer(dtime)
+	hunger_timer = hunger_timer + dtime
+	health_timer = health_timer + dtime
+	action_timer = action_timer + dtime
+
+	if action_timer > HUNGER_MOVE_TICK then
```

Comment:
This function likely needs a full rewrite to optimize. However, I don't think I'll be able to remove the globalstep part entirely, mostly because the proper replacement using ```minetest.after``` would cause lots of timers to fire per player, and you'd end up with more timer pressure from that.